### PR TITLE
Add interpreter for WhileOp

### LINKED ISSUE #992: Add interpreter for while
Following the [testing guidelines](https://github.com/openxla/stablehlo/blob/main/docs/interpreter_checklist.md)

### PR DESCRIPTION
**Pull request by ghpvnist**
We have the following constraints in the spec:

```
(I1) operand: variadic number of tensors or tokens
(I2) cond: function
(I3) body: function
(C1) `cond` has type `(T0, ..., TN-1) -> tensor<i1>`, where
     `Ti` = `type(operand[i])`.
(C2) `body` has type `(T0, ..., TN-1) -> (T0, ..., TN-1)`, where
     `Ti` = `type(operand[i])`.
(C3) For all `i`, `type(results[i])` = `type(operand[i])`.
```

These constraints will be comprehensively covered by the following tests:

```
I1: a) operand is not variadic tensors or tokens. (Covered by ODS).
I2: a) cond is not a function. (Covered by ODS).
I3: a) body is not a function. (Covered by ODS).
C1: a) `cond` does not have type `(T0, ..., TN-1) -> tensor<i1>`, where
     `Ti` = `type(operand[i])`.
C2: a) `body` does not have type `(T0, ..., TN-1) -> (T0, ..., TN-1)`, where
     `Ti` = `type(operand[i])`.
C3: a) For any `i`, `type(results[i])` != `type(operand[i])`.
```

If we drop the "Covered by ODS" pieces, this will leave us with the following test cases:

```
C1a: `cond` does not have type `(T0, ..., TN-1) -> tensor<i1>`, where
     `Ti` = `type(operand[i])`.
C2a: `body` does not have type `(T0, ..., TN-1) -> (T0, ..., TN-1)`, where
     `Ti` = `type(operand[i])`.
C3a: For any `i`, `type(results[i])` != `type(operand[i])`.
```

closes #992

### CODE REVIEW

**Comment by ghpvnist** [comment]
This PR also contains temporary commit rebased from #1199. You may ignore the following files for now, and partially review the changes related to WhileOp for others:
`stablehlo/reference/`:
* Element.cpp
* Element.h
* Ops.h

`stablehlo/tests/`:
* interpret_compare.mlir
* print_stablehlo.mlir
* stablehlo_legalize_to_vhlo.mlir

File: stablehlo/tests/ops_stablehlo.mlir [hunk 1]
```diff
@@ -2897,6 +2876,114 @@ func.func @reverse_i2(%operand: tensor<3x2xi32>) -> tensor<3x2xi32> {
 
 // -----
 
+// CHECK-LABEL: func @while
```
Comment (sdasgup3) [inline stablehlo/tests/ops_stablehlo.mlir:2879]: Just curious to know, if these tests duplicate of the ones added in verify_while.mlir?
In any case, we can keep  them in its dedicated file `verify_while.mlir`?

File: stablehlo/tests/infer_stablehlo.mlir [hunk 2]
```diff
@@ -592,6 +570,51 @@ func.func @get_dimension_size(%arg0: tensor<4x2xf32>) -> tensor<index> {
...
+    "stablehlo.return"(%3, %arg10, %arg11) : (tensor<i32>, tensor<i32>, tensor<i32>) -> ()
+  }) : (tensor<i32>, tensor<i32>, tensor<i32>) -> (tensor<i32>, tensor<i32>, tensor<i32>)
+  // CHECK: types0 = tensor<i32>, types1 = tensor<i32>, types2 = tensor<i32>
+  %4 = "hlo_test_infer.get_return_types"(%1#0) : (tensor<i32>) -> tensor<index>
+  func.return %4 : tensor<index>
+}
+
+// -----
+
+func.func @while_c3(%arg0: tensor<3xf32>) -> tensor<3xf32> {
```
Comment (sdasgup3) [inline stablehlo/tests/infer_stablehlo.mlir:595]: Is this an additional test?

File: docs/spec.md [hunk 3]
```diff
@@ -5453,24 +5457,28 @@ The behavior of an infinite loop is TBD
...
-%results0, %results1 = "stablehlo.while"(%operand0, %operand1) ({
-  ^bb0(%arg0: tensor<i32>, %arg1: tensor<i32>):
-    %0 = "stablehlo.compare"(%arg0, %arg1) {
+// %init_i: 1
+// %init_sum: 0
+// %one: 1
+// %ten: 10
+%results0, %results1 = "stablehlo.while"(%init_i, %init_sum) ({
+  ^bb0(%curr_i: tensor<i64>, %curr_sum: tensor<i64>):
+    %cond = "stablehlo.compare"(%curr_i, %ten) {
```
Comment (sdasgup3) [inline docs/spec.md:5466]: We have been consistently using  `%arg0, %arg1` for the arguments. Can we use that here as well?

**Review by sdasgup3 — COMMENTED** [review]
a few minor comments. I have not reviewed the code related to compare op. I hope this PR will be merged after the `compareOp` interpreter lands.

File: stablehlo/tests/infer_stablehlo.mlir (hunk 2 above)
Comment (ghpvnist) [reply stablehlo/tests/infer_stablehlo.mlir:595]: No, this is moved from `verify_while.mlir`

**Review by sdasgup3 — APPROVED** [review]

File: stablehlo/dialect/StablehloOps.td [hunk 4]
```diff
@@ -1239,23 +1239,22 @@ def StableHLO_WhileOp: StableHLO_Op<"while", [
     https://github.com/openxla/stablehlo/blob/main/docs/spec.md#while
 
     Example:
-    ```mlir
```
Comment (burmako) [inline stablehlo/dialect/StablehloOps.td:1242]: We still want the "```mlir" part, right?

File: stablehlo/tests/interpret_while.mlir [hunk 5]
```diff
@@ -1,26 +1,26 @@
...
-  // int i = 10;
-  // int state = 0;
-  // while (i > 0) {
-  //   i -= 1;
-  //   state += 1;
+  // int i = 0;
+  // int sum = 0;
+  // while (i < 10) {
+  //   sum += 1;
+  //   i += ;
```
Comment (burmako) [inline stablehlo/tests/interpret_while.mlir:8]: "i += 1"?

File: stablehlo/tests/infer_stablehlo.mlir [hunk 6]
```diff
@@ -592,6 +570,51 @@ func.func @get_dimension_size(%arg0: tensor<4x2xf32>) -> tensor<index> {
...
+  %cst_0 = arith.constant dense<0> : tensor<1xi32>
+  %cst_1 = arith.constant dense<[100, 100]> : tensor<2xi32>
+  %cst_2 = arith.constant dense<1.00> : tensor<1xf32>
+  // expected-error @+1 {{inferred type(s) 'tensor<1xi32>', 'tensor<2xi32>', 'tensor<1xf32>', 'tensor<3xf32>' are incompatible with return type(s) of operation 'tensor<1xi32>', 'tensor<2xi32>', 'tensor<3xf32>', 'tensor<1xf32>'}}
```
Comment (burmako) [inline stablehlo/tests/infer_stablehlo.mlir:599]: I don't think this test is testing anything new in comparison with the previous one?

**Review by burmako — CHANGES_REQUESTED** [review]

File: stablehlo/tests/infer_stablehlo.mlir (hunk 6 above)
Comment (ghpvnist) [reply stablehlo/tests/infer_stablehlo.mlir:599]: Agreed, the checklist does not require a negative test to be added for every constraint.

**Review by burmako — APPROVED** [review]

**Pull request merged**